TrueCraft/QueryProtocol.cs

```diff
@@ -0,0 +1,231 @@
+            response.Write((byte)0x00);
+
+            SendResponse(response, clientEP);
+        }
+
+        private void CheckVersion(BinaryReader stream)
+        {
+            byte ver1 = stream.ReadByte();
+            byte ver2 = stream.ReadByte();
+            if (ver1 != ProtocolVersion.Item1 || ver2 != ProtocolVersion.Item2) 
+                throw new Exception("Incorrect Protocol Version");
+        }
+        private void CheckType(byte Type, BinaryReader stream)
+        {
+            byte type = stream.ReadByte();
+            if (type != Type) throw new Exception("Incorrect Type");
+        }
+        private void CheckHead(byte Type, BinaryReader stream)
+        {
+            CheckVersion(stream);
+            CheckType(Type, stream);
+        }
+        private int GetSessionId(BinaryReader stream)
+        {
+            stream.BaseStream.Position = 3;
+            return stream.ReadInt32();
+        }
+        private int GetToken(BinaryReader stream)
+        {
+            stream.BaseStream.Position = 7;
+            return stream.ReadInt32();
+        }
+        private BinaryReader GetStream(byte[] buffer)
+        { return new BinaryReader(new MemoryStream(buffer)); }
+        private BinaryWriter GetStream()
+        { return new BinaryWriter(new MemoryStream()); }
+        private void WriteHead(byte type, QueryUser user, BinaryWriter stream)
+        {
+            stream.Write(type);
+            stream.Write(user.SessionId);
+        }
+        private void SendResponse(BinaryWriter res, IPEndPoint destination)
+        {
+            byte[] data = ((MemoryStream)res.BaseStream).ToArray();
+            Udp.Send(data, data.Length, destination);
+        }
+        private QueryUser GetUser(IPEndPoint ep)
+        {
+            QueryUser user;
+            lock (UserLock)
+                if (!UserList.TryGetValue(ep, out user)) throw new Exception("Undefined user");
+            return user;
+        }
+        private Dictionary<string, string> GetStats()
+        {
+            var stats = new Dictionary<string, string>();
+            stats.Add("hostname", Program.ServerConfiguration.MOTD);
+            stats.Add("gametype", "SMP");
+            stats.Add("game_id", "TRUECRAFT");
+            stats.Add("version", "1.0");
+            stats.Add("plugins", "");
+            stats.Add("map", Program.Server.Worlds.First().Name);
+            stats.Add("numplayers", Program.Server.Clients.Count.ToString());
+            stats.Add("maxplayers", "64");
+            stats.Add("hostport", Program.ServerConfiguration.ServerPort.ToString());
+            stats.Add("hostip", Program.ServerConfiguration.ServerAddress);
+            return stats;
+        }
+        private List<string> GetPlayers()
+        {
+            var names = new List<string>();
+            lock (Program.Server.ClientLock)
+                foreach (var client in Program.Server.Clients)
```

Comment:
Instead of using `Program.Server`, how about passing an instance of `IMultiplayerServer` to the constructor for this class?